src/lowering/desugar.ml

```diff
@@ -68,6 +68,20 @@ and exp' at note = function
     I.PrimE (I.RelPrim (!ot, o), [exp e1; exp e2])
   | S.ShowE (ot, e) ->
     I.PrimE (I.ShowPrim !ot, [exp e])
+  | S.ToCandidE (ot, es) ->
+    begin match !ot with
+    | T.Tup ts1 ->
+      I.PrimE (I.SerializePrim [T.Tup ts1], [tupE (exps es)])
+    | _ -> assert false
+    end
+  | S.FromCandidE (ot, e) ->
+    begin match !ot with
+    | T.Opt (T.Tup ts1) ->
+      I.PrimE (I.DeserializePrim ts1, [exp e])
+    | T.Opt t ->
+      I.PrimE (I.DeserializePrim [t], [exp e])
+    | _ -> assert false
+    end
```

Comment:
```suggestion
  | S.FromCandidE (ot, e) ->
    begin match T.normalize (!ot) with
    | T.Opt t1 -> 
      (match T.normalize t1 with
        | (T.Tup ts1) ->
          I.PrimE (I.DeserializePrim ts1, [exp e])
        | t1 -> I.PrimE (I.DeserializePrim [t1], [exp e]))
    | _ -> assert false
    end
```
(and similarly in ToCandidE unless you know that will already be a normalized T.Tup).